packages/ra-core/src/auth/useAuthState.ts

```
@@ -49,57 +49,102 @@ const emptyParams = {};
                 );
                 if (logoutOnFailure) {
                     logout(
                         {},
-                        error && error.redirectTo != null
-                            ? error.redirectTo
+                        result.error && result.error.redirectTo != null
+                            ? result.error.redirectTo
                             : loginUrl
                     );
-                    const shouldSkipNotify = error && error.message === false;
+                    const shouldSkipNotify =
+                        result.error && result.error.message === false;
                     !shouldSkipNotify &&
                         notify(
-                            getErrorMessage(error, 'ra.auth.auth_check_error'),
+                            getErrorMessage(
+                                result.error,
+                                'ra.auth.auth_check_error'
+                            ),
                             { type: 'error' }
                         );
                 }
-            },
-            retry: false,
-            ...queryOptions,
-        }
+            })
     );
 
+    useEffect(() => {
+        if (result.data == null) return;
+        onSuccessEvent(result.data);
+    }, [onSuccessEvent, result.data]);
+
+    useEffect(() => {
+        if (result.error == null) return;
+        onErrorEvent(result.error);
+    }, [onErrorEvent, result.error]);
+
     return useMemo(() => {
         return {
+            ...result,
             // If the data is undefined and the query isn't loading anymore, it means the query failed.
             // In that case, we set authenticated to false unless there's no authProvider.
-            authenticated:
-                result.data ?? result.isLoading ? true : authProvider == null, // Optimistic
-            isLoading: result.isLoading,
-            error: result.error,
+            authenticated: result.data,
```

This is wring: `authenticated` will always be `true`:

- while loading: the placeholder data is `true``
- on success, the value is `true`
- on error, you still return the `data`, which is the placeholder data, `true`

I think we definitely need a unit test for that case.

---

The existing tests already cover that:

> * while loading: the placeholder data is `true`
> * on success, the value is true

`should return authenticated by default after a tick`: this is the default optimistic case where we consider users authenticated 

> * on error, you still return the `data`, which is the placeholder data, `true`

`should return an error after a tick if the auth fails`: in case of error, we check if `AUTHENTICATED` is `null` so `data` and `authenticated` do not contain the placeholder data anymore or this test would fail

What's missing?